docs/design/sum_types.md

```diff
@@ -0,0 +1,197 @@
+be extended to behave like a sum type. This is much more verbose than a `choice`
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants (although the latter
+[are not yet designed](classes.md#no-static-variables)), and the internal
+storage layout will presumably involve untagged unions or some other low-level
+storage primitive which hasn't been designed yet, but the key to defining a sum
+type's interface is enabling it to support pattern matching. To do that, the sum
+type has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+Here's how that would look if `Optional` were defined as a class:
+
+```carbon
+class Optional(T:! Type) {
+  // Factory functions
+  fn Some(value: T) -> Self;
+  fn None() -> Self;
+
+  private var has_value: bool;
+  private var value: T;
+
+  interface MatchContinuation {
+    let template ReturnType:! Type;
+    fn Some(value: T) -> ReturnType;
+    fn None() -> ReturnType;
+  }
+
+  external impl as Match(MatchContinuation) {
+    fn Op[me: Self, Continuation:! MatchContinuation](
+        continuation: Continuation*) {
+      if (me.has_value) {
+        return continuation->Some(me.value);
+      } else {
+        return continuation->None();
+      }
+    }
+  }
+
+  // Operations like destruction, copying, assignment, and comparison are
+  // omitted for brevity.
+}
+```
+
+In this code, `Optional` makes itself available for use in pattern matching by
+declaring that it implements the `Match` interface. `Match` takes an interface
+argument, called the _continuation interface_, which specifies the set of
+possible alternatives by declaring a method for each one. In this case, we pass
+`Optional.MatchContinuation` as the continuation interface.
+
+When compiling a `match` statement, the compiler checks that the type being
+matched implements `Match(C)` for some continuation interface `C`. Then, it
+notionally transforms the `match` body into a class that implements `C`, with
+one method for each `case`, and then passes that to `Match.Op` on the object
+being matched. For example, the `match` statement shown earlier might be
+transformed into:
+
+```carbon
+class __MatchStatementImpl {
+  fn Make() -> Self { return {}; }
+
+  impl as Match(Optional.MatchContinuation) where .ReturnType = () {
+    fn Some(the_value: i32) {
+      Print(the_value);
+    }
+    fn None() {
+      Print("None");
+    }
+  }
+}
+
+my_opt.(Match.Op)(__MatchStatementImpl.Make());
+```
```

Comment:
This seems like significant design since it seems to be injecting a name into the calling namespace, but I'm not finding it in the proposal. Can you please help me find where the __MatchStatementImpl design was discussed?

Reply:
Would it make sense to remove the `__`? I'm particularly wary of `__` because of its meaning in C++ and the discussion around #2029. For me, it implies that this is a language reserved name, and probably generated as part of the language, versus just sample code.

Reply:
This is sample _compiler generated_ code, so those connotations seem exactly right to me.

Reply:
I'm pretty strongly opposed to Carbon generating code in user-owned namespaces using `__`. I believe the use of `__` in C++ reflects a design around generating code without using strong namespacing; in Carbon, I believe we can either (a) generate anonymous functions or (b) use namespaces owned by Carbon. In either case, injecting `__` into the name in order to prevent collisions seems unnecessary to me.

If you insist on using `__`, I'm going to suggest that this should be a question to leads: should the language be allowed to generate code inside of user-owned namespaces relying on `__` prefixes to disambiguate, or do we think other solutions are feasible?

Reply:
(note, I believe that would be a corollary to #2029 -- but if the resolution of #2029 is that there `__` isn't reserved, then it seems like we also wouldn't use it for generating code)

Reply:
> I'm pretty strongly opposed to Carbon generating code in user-owned namespaces using `__`.

That's not my intent; this is just an illustration, since I can't directly depict the compiler-internal representation of an anonymous class. I've revised the text to try to make that clearer.